File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        x = [1,-1,0,0]
        y = [0,0,1,-1]
        def dfs(i,j):
            grid[i][j] = "0"
            for k in range(4):
                nexti = i + x[k]
                nextj = j + y[k]
                if nexti >= 0 and nexti < len(grid) and nextj >= 0 and nextj < len(grid[0]) and grid[nexti][nextj] == "1":
                    dfs(nexti,nextj)
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1":
                    count+=1
                    dfs(i,j)
        return count
        
        
        
        

File: 200-number-of-islands/number-of-islands.java
class Solution {
    public boolean visited [][];
    public char gridG[][];
    public int [] X = new int[]{1,-1,0,0};
    public int [] Y = new int[]{0,0,1,-1};
    public int numIslands(char[][] grid) {
        visited = new boolean[grid.length][grid[0].length];
        gridG = grid;
        int count = 0;
        for(int i = 0;i<grid.length;i++){
            for(int j = 0;j<grid[0].length;j++){
                if(grid[i][j] == '1' && visited[i][j] == false){
                    count+=1;
                    dfs(i,j);
                }
            }
        }
        return count;
    }
    public void dfs(int i,int j){
        visited[i][j] = true;
        for(int k = 0;k<4;k++){
            int nexti = i + X[k];
            int nextj = j + Y[k];
            if(nexti < gridG.length && nextj < gridG[0].length && nexti >= 0 && nextj >= 0){
                if(gridG[nexti][nextj] == '1' && visited[nexti][nextj] == false){
                    dfs(nexti,nextj);
                }
            }
        }
    }
}